src/shims/unix/linux/epoll.rs

```diff
@@ -500,9 +578,21 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
                         let event_instance =
                             EpollEventInstance::new(flags, epoll_event_interest.data);
                         ready_list.insert(epoll_key, event_instance);
+                        // Collect all thread id of blocked threads that received notification.
+                        let thread_id = epoll_event_interest.thread_id;
+                        if this.has_blocked_on_epoll(thread_id) {
+                            waiters.push(thread_id);
```

Comment:
could also just wake here and record in a hashset whether the thread id was already woken

Reply:
For the new change in https://github.com/rust-lang/miri/pull/3804/commits/9cb087dc8b529266bebcc9637d9ff8993c88c811, I was trying to avoid using ``this.unblock_thread`` inside 
```rust
if let Some(epoll_interests) = this.machine.epoll_interests.get_epoll_interest(id) { }
```
because it will trigger this error: 
```bash
error[E0502]: cannot borrow `*this` as mutable because it is also borrowed as immutable
   --> src/shims/unix/linux/epoll.rs:594:29
    |
568 |         if let Some(epoll_interests) = this.machine.epoll_interests.get_epoll_interest(id) {
    |                                        ---------------------------- immutable borrow occurs here
...
573 |             for weak_epoll_interest in epoll_interests {
    |                                        --------------- immutable borrow later used here
...
594 |                             this.unblock_thread(thread_id, BlockReason::Epoll)?;
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```
So instead I just store the ThreadIds that I want to unblock into a Vec then unblock them altogether outside of the block.